docs/StyleGuide.md

```diff
@@ -0,0 +1,1142 @@
+   Not doing so makes it harder to read and is sloppy.
+
+   ```haskell
+   avg x y = (x + y) / 2
+
+   let ((x, y), z) = foo
+   in (y, z)
+   ```
+
+   The only exception is in tuple sections:
+
+   ```haskell
+   (,) <$> foo <*> bar
+   (True,) <$> foo
+   ```
+
+6. __Function composition and the dollar operator__:
+
+   Choose between using parenthesis, `$` and `.` in whichever way you think
+   results in the most readable code.
+
+7. __Blank lines__
+   Keep two empty lines between top level definitions.
+
+   This is natural and useful especially when one is using a single empty
+   lines to delimit code sections inside a longer code block.
+
+   ```haskell
+   data Foo = Foo
+
+
+   foo :: Foo -> ()
+   foo Foo = ()
+   ```
+
+   Use this wisely, sometimes it makes sense to use a single empty line, for
+   example to delimit short instances which all refer to the same data type.
+
+   Sometimes it's more logical too keep `Foo` and `foo` close together but
+   `Bar` more distinct, e.g.
+
+   ```haskell
+   data Foo = Foo
+
+   foo :: Foo -> ()
+   foo Foo = ()
+
+  
+   data Bar = Bar
+
+   bar :: Bar -> ()
+   bar Bar = ()
+   ```
+
+   Always end a file with a *newline*, which is not the same as a blank line.
+   ```
+   -- NO
+   ..
+
+   <EOF>
+
+   -- NO
+   ..<EOF>
+
+   -- YES
+   ..
+   <EOF>
+   ```
+   *Why:* see [this StackOverflow answer][posix-line], moreover, GitHub will
+   highlight a missing newline at the end of the file.
+
+   [posix-line]: https://stackoverflow.com/questions/729692/why-should-text-files-end-with-a-newline#answer-729795
+
+8. __Sections__: we group related definitions in sections that start with a
+   section title. The same grouping can be replicated in the export list.
+
+    ```haskell
+    module AmazingModule
+      ( -- Foo
+        Foo (..)
+      , mkFoo
+        -- Bar
+      , ..
+      ) where
+
+    --
+    --  Foo
+    --
+
+    data Foo = ..
+
+    mkFoo :: ..
+
+    ..
+
+    --
+    -- Bar
+    --
+    -- Bar is bla bla
+    --
+
+    type Bar = ..
+    ..
+    ```
+
+    The two lines of the section header are each 80 characters in total. The
+    title is indented by two spaces. The section header can contain more text,
+    which is separated from the first line by one empty comment line. The
+    section header has a single blank line above and below it.
+
+9. __Comment style__: in general we tend to use `--` instead of `{- .. -}`. We
+   sometimes make exceptions for big non-Haddock comments.
+
+10. __Haddock formatting__: we use [Haddock formatting][haddock-formatting] in
+    docstrings. We also do this in comments for consistency.
+
+    ```haskell
+    -- | Short title
+    --
+    -- Longer description .. 'Foo' .. "Data.String" .. @a@ .. /not/ ..
+    -- __never__ .. called \"foo bars\" .. alternative style " foo bars "
+    -- .. @'Foo' a@
+    --
+    -- > foo bar baz
+    --
+    -- .. ends here.
```

Comment:
Don't we usually have an empty comment line between the haddock/description and the function type signature like this:

```haskell
-- | Description ....
-- 
function :: A -> B
function = ...
```

Reply:
Resolved.